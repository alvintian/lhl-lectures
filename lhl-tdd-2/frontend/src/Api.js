export const fetchBeers = () => {
  return new Promise((resolve, reject) => {
    fetch("/api/beers")
      .then(resp => resp.json())
      .then(data => {
        resolve(data);
      })
      .catch(() => {
        console.error("api request failed");
      });
  });
};

export const createBeer = async beerAttributes => {
  try {
    const response = await fetch("/api/beers", {
      method: "POST",
      body: JSON.stringify({ beer: beerAttributes })
    });
    return response.json();
  } catch (err) {
    throw err;
  }
};

export default {
  fetchBeers
};
