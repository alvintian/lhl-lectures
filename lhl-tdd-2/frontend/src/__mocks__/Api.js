export const beerData = [
  {
    id: 1,
    name: "Lone Pine",
    style: "IPA"
  }
];

export const fetchBeers = jest.fn(() => Promise.resolve(beerData));
