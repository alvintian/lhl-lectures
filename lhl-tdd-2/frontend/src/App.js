import React, { Component } from "react";
import BeerTable from "./BeerTable";
import { fetchBeers } from "./Api";

import NewBeerForm from "./NewBeerForm";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = { beers: [], showForm: false };
  }
  componentDidMount() {
    fetchBeers().then(({ beers }) => {
      this.setState({ beers });
    });
  }

  handleShowForm = event => {
    event.preventDefault();
    this.setState({ showForm: true });
  };

  handleNewBeer = data => {};

  render() {
    return (
      <div>
        <h1>Beer App</h1>
        <a href="#" onClick={this.handleShowForm}>
          New Beer
        </a>
        {this.state.showForm && <NewBeerForm onSubmit={this.handleNewBeer} />}
        <BeerTable beers={this.state.beers} />
      </div>
    );
  }
}
export default App;
