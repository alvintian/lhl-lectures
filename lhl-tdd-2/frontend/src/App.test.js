import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import "./testSetup";
import { shallow } from "enzyme";

import { fetchBeers } from "./Api";
import { beerData } from "./__mocks__/Api";

jest.mock("./Api");

describe("<App/>", () => {
  describe(".render()", () => {
    it("renders the component correctly based of off the default state", () => {
      const wrapper = shallow(<App />);
      expect(wrapper).toMatchSnapshot();
    });

    it("renders the beer table when beers exist on the state", () => {
      const wrapper = shallow(<App />);
      wrapper.setState({ beers: beerData });
      expect(wrapper).toMatchSnapshot();
    });

    it("renders the <NewBeerForm/> if showForm === true", () => {
      const wrapper = shallow(<App />);
      wrapper.setState({ showForm: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe(".componentDidMount()", () => {
    it("calls fetchBeers()", async () => {
      const wrapper = shallow(<App />);
      expect(fetchBeers).toHaveBeenCalled();
    });
  });
});
