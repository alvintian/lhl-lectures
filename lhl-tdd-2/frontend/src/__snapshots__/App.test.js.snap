// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> .render() renders the <NewBeerForm/> if showForm === true 1`] = `
<div>
  <h1>
    Beer App
  </h1>
  <a
    href="#"
    onClick={[Function]}
  >
    New Beer
  </a>
  <NewBeerForm
    onSubmit={[Function]}
  />
  <BeerTable
    beers={Array []}
  />
</div>
`;

exports[`<App/> .render() renders the beer table when beers exist on the state 1`] = `
<div>
  <h1>
    Beer App
  </h1>
  <a
    href="#"
    onClick={[Function]}
  >
    New Beer
  </a>
  <BeerTable
    beers={
      Array [
        Object {
          "id": 1,
          "name": "Lone Pine",
          "style": "IPA",
        },
      ]
    }
  />
</div>
`;

exports[`<App/> .render() renders the component correctly based of off the default state 1`] = `
<div>
  <h1>
    Beer App
  </h1>
  <a
    href="#"
    onClick={[Function]}
  >
    New Beer
  </a>
  <BeerTable
    beers={Array []}
  />
</div>
`;
